classes/range.js

```diff
@@ -190,6 +190,35 @@ class Range {
     return false
   }
 }
+
+function buildMemoKeyFromOptions(options) {
```

Comment:
You may also want to try a handwritten key such as:

```
`${options.includePrerelease ? 1 : 0},${options.loose ? 1 : 0}...
```

Which is what we do in the TypeScript compiler. Not sure if it's faster than here but could be interesting to test.

Reply:
It doesn't hurt the performance:

```md
buildMemoKeyFromOptions({"includePrelease":true}) x 1,121,468,719 ops/sec ±0.63% (96 runs sampled)
buildMemoKeyFromOptions({"includePrelease":true,"loose":true}) x 1,132,354,133 ops/sec ±0.67% (95 runs sampled)
buildMemoKeyFromOptions({"includePrelease":true,"loose":true,"rtl":true}) x 1,122,788,698 ops/sec ±0.53% (96 runs sampled)
```

After 1B ops/s, it almost has no effect choosing one or another.
The only drawback is increasing the lru-cache key by 6 characters instead of having it as 1.

If NPM teams prefer readability over a tiny cache key, I can change the implementation.

Reply:
This is actually a good use case for a bit flag.  It's a bunch of booleans, and bitwise operations are super fast.  And while bit ops are often frowned upon as being dense or obscure, if organized properly, they're much more maintainable than a series of `if/else` statements and hoping that you captured each case.

```js
let s = 1
const FLAG_includePrerelease = s
s<<=1
const FLAG_loose = s
s<<=1
const FLAG_rtl = s
s<<=1
// ...

function buildMemoKeyFromOptions(options) {
  // now just bitwise-OR them all together as appropriate
  return (
    options.loose ? FLAG_loose : 0
  ) | (
    options.rtl ? FLAG_rtl : 0
  ) | (
    options.includePrerelease ? FLAG_includePrerelease : 0
  )
}
```

Reply:
@isaacs Options are part of the API? If not, I can refactor to use bit flags and see what is the perf output of it.

I didn't want to change that before because options could be used by users and I didn't want to introduce any breaking change.

Reply:
I don't think the suggestion is to change Options, but to use a different way of constructing a key.

That being said, I think it should be possible to internally store flags for these and then recustruct the options bag when asked on whichever methods that request it. Then, internally, things can use the fast check and pass `number` around. Spitballing, though.

Reply:
Yes, I realized a few seconds after writing that, and yes, using the option internally and then exposing makes sense, I'll try using the getter to compute the options and internally using just the bit flag.

Reply:
Thinking further, the user already has to pass the object in, so you probably can just keep that one and not reconstruct it.

Reply:
@jakebailey I tried, but it would break a lot of tests if I did.

I put the version of parseOptions with bit flags here: https://github.com/h4ad-forks/node-semver/commit/6c7a68ad9849cadbc624ef5208662ae1ea01d846

@isaacs Take a look and see if this PR is worth pushing, I made a lot of changes.

Also, I introduce 1 small breaking change, now the numbers passed to `parseOptions` are considered valid options, so instead of re-parsing it by passing the object, I use many of the options as bit flags to construct new objects that are inside the functions.

About performance, no performance penalties and I assume it got a bit faster because I now parse the options as an object once and then just parse the options again as a number which saves some comparisons.

Reply:
The only downside I notice when using bitmask is the memory usage (ref: https://github.com/npm/node-semver/pull/528#discussion_r1155041551) is the same as allocating the objects at runtime (8.7mb).

I don't know why this behavior is happening, but I just want to point out that it's not as memory efficient as Object.freeze.